### PR TITLE
feat: Advanced Segment Anything Model

### PR DESCRIPTION
**Pull request by shondle**
https://github.com/shondle/label-studio-ml-backend/assets/106922533/42a8a535-167c-404a-96bd-c2e2382df99a
> Video demo explained [here](#video-example-explained)

New SAM update. Lots of new features, and added a new model for higher accessibility for people without access to higher-end compute resources to load SAM.

This adds a new prediction backend I am calling Advanced SAM. This allows many more options for adjusting annotation predictions than before with the ONNX models, and includes [a video tutorial](https://drive.google.com/file/d/1OMV1qLHc0yYRachPPb8et7dUBjxUsmR1/view?usp=sharing) to help those getting started. 

## Advanced SAM Features

*Adjusting Regions With Multiple Input Prompts At A Time*
- :fire:  Combine multiple keypoints and a rectangle label together to fine-adjust areas of region predictions. 
- :fire: Combine negative keypoints with other smart tools to tell the model where to avoid when generating predictions.
- :fire: Unlimited keypoints allowed for generating predictions for each region, allowing ultimate customizability and time saved for annotators!
- :fire: Move around the current smart keypoint or rectangle label for the current region and watch as predictions are adjusted real-time!

*Model Predictions*
- :fire: Choose to use the SAM architecture from Facebook or run a lighter version of SAM from the MobileSAM paper. I am able to run this **and receive predictions :boom: in <1s on my Laptop GPU (RTX 3060)!**:boom: I record the entire video tutorial on my laptop while gathering predictions real-time.

## ONNX SAM Features

*Features*
- Added ability to use smart rectangle labels to generate predictions

*Issues Addressed*
- ONNX image predictions offset issue addressed in README. This was due to the need to specify image size before generating the model. #278 
- Video tutorial to show how to use the backend (should address some of the no kwargs received issues). #287 

## Demo Explanation
- In the video above, when I click the turtle's head, it accidentally selects the reflection of the head in the water as well.
    - So, I use a negative keypoint tool and place it in the water.
    - As such, SAM batches together the positive keypoint used to select the head of the turtle and the negative keypoint in the water to provide a prediction that just surrounds the turtle's head.

- At the end of the video of the car, I try to select the road trailing the car.
  - First, I place a positive keypoint on the road. SAM predicts part of the sidewalk to be a part of the road.
  - So, I use a negative keypoint on the sidewalk. This keeps the prediction on the road now, but leaves the bottom left corner without an annotation.
  - As such, I use a rectangle label along this part of the road. This batches together the first positive keypoint, the negative keypoint on the sidewalk, and the rectangle label to form a precise prediction that covers the road trailing the car.

## TL;DR
:punch: Use the AdvancedSAM backend to batch together a rectangle label, and unlimited positive and negative keypoints to fine-adjust model predictions and see model predictions change real-time while annotating. Use the MobileSAM architecture to run predictions on consumer hardware. Video tutorial included to help people get started. :punch:

### CODE REVIEW

**Comment by hogepodge** [comment]
This LGTM, and it's a great update to the model. Thanks @shondle!

File: label_studio_ml/examples/segment_anything_model/advanced_sam.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+                        input_box = input_box[None, :]
+        except Exception as e:
+            print(e)
+            print("No previous tasks")
+        
+        return points, labels, input_box, alias, box_width, box_height
+
+    def set_image(self, predictor, image_path):
+        # this is to speed up inference after the first time you selected an image
+        global PREV_IMG_PATH
```
Comment (niklub) [inline label_studio_ml/examples/segment_anything_model/advanced_sam.py:156]: I'm not sure this gonna work with concurrent requests

File: label_studio_ml/examples/segment_anything_model/advanced_sam.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        predictions.append({"result": results,
+                            "model_version": "vit_h"
+        })
+
+        return predictions
+
+
+
+
+    def fit(self, completions, workdir=None, **kwargs):
```
Comment (niklub) [inline label_studio_ml/examples/segment_anything_model/advanced_sam.py:252]: Could you please remove this function? The interface was reworked in the latest base branch and no longer compatible with this signature

**Review by niklub — CHANGES_REQUESTED** [review]

**Comment by niklub** [comment]
That's excellent @shondle ! Let's slightly modify the code to make it fully compatible with the new ML backend codebase, I'll follow up with other changes to be made quite soon. Thank you!

File: label_studio_ml/examples/segment_anything_model/advanced_sam.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+import string
+import random
+import torch
+import json
+
+
+SAM_CHOICE = os.environ.get("SAM_CHOICE", "MobileSAM") # other option is just SAM
+
+if SAM_CHOICE == "MobileSAM":
+    from mobile_sam import SamPredictor, sam_model_registry
```
Comment (botcs) [inline label_studio_ml/examples/segment_anything_model/advanced_sam.py:18]: This file is missing at the moment from the PR and the original version

**Comment by hogepodge** [comment]
@niklub @shondle any updates on the progress of this? We've got a lot of interest, and I'd like to move forward on getting this merged

File: label_studio_ml/examples/segment_anything_model/Dockerfile
Comment (hogepodge) [inline label_studio_ml/examples/segment_anything_model/Dockerfile:1]: I'd like for a few things to be adjusted. First, a secret is being leaked into the Dockerfile in an environment variable. In my own testing, this actually does nothing. However, with some minor modifications it's possible to use Label Studio as an image source. If the user sets the variables:

```
ACCESS_TOKEN=<token>
LABEL_STUDIO_HOST=<ls_host_ip>
```

Loads those environment variables into the model:

```
ACCESS_TOKEN = os.environ.get("ACCESS_TOKEN", None)
LABEL_STUDIO_HOST = os.environ.get("LABEL_STUDIO_HOST", None)
```

and updates the load image code

```
from label_studio_tools.core.utils.io import get_local_path
...
image_path = get_local_path(img_path, access_token=ACCESS_TOKEN, hostname=LABEL_STUDIO_HOST)
```

Label Studio can be used to host the images.

**Comment by niklub** [comment]
> @niklub @shondle any updates on the progress of this? We've got a lot of interest, and I'd like to move forward on getting this merged

I'm merging it to the base branch now to continue adjusting with the new ML backend interface. @shondle @hogepodge I'll let you know when it is done and your review is needed
